massa-final-state/src/final_state.rs

```diff
@@ -235,12 +189,96 @@ impl FinalState {
+                self.pos_state
+                    .cycle_history
+                    .back_mut()
+                    .ok_or(FinalStateError::SnapshotError(String::from(
+                        "Invalid cycle_history",
+                    )))?;
+
+            for _ in current_slot.period..end_slot.period {
+                latest_cycle_info.rng_seed.push(false);
+            }
+        } else {
+            // Here, we we also complete the cycle_infos in between
+
+            // Firstly, complete the first cycle
+            let latest_cycle_info =
+                self.pos_state
+                    .cycle_history
+                    .back_mut()
+                    .ok_or(FinalStateError::SnapshotError(String::from(
+                        "Invalid cycle_history",
+                    )))?;
+
+            for _ in current_slot.period..self.config.periods_per_cycle {
+                latest_cycle_info.rng_seed.push(false);
+            }
+            latest_cycle_info.complete = true;
+
+            // feed final_state_hash to the completed cycle
+            self.pos_state
+                .feed_cycle_state_hash(current_slot_cycle, self.final_state_hash);
+
+            self.pos_state
+                .feed_selector(current_slot_cycle.checked_add(2).ok_or_else(|| {
+                    FinalStateError::PosError("cycle overflow when feeding selector".into())
+                })?)
+                .map_err(|_| {
+                    FinalStateError::PosError("cycle overflow when feeding selector".into())
+                })?;
+
+            // Then, build all the already completed cycles
+            for cycle in current_slot_cycle + 1..end_slot_cycle {
+                // TODO: build a new cycle by repeating the one before. How do we handle this if latest cycle info is not complete?
+                let last_slot = Slot::new_last_of_cycle(
+                    cycle,
+                    self.config.periods_per_cycle,
+                    self.config.thread_count,
+                )
+                .map_err(|err| {
+                    FinalStateError::InvalidSlot(format!(
+                        "Cannot create slot for interpolating downtime: {}",
+                        err
+                    ))
+                })?;
+
+                self.pos_state
+                    .create_new_cycle_from_last(&latest_consistent_cycle_info, last_slot)
+                    .map_err(|err| FinalStateError::PosError(format!("{}", err)))?;
+
+                self.pos_state
+                    .feed_cycle_state_hash(cycle, self.final_state_hash);
+
+                self.pos_state
+                    .feed_selector(cycle.checked_add(2).ok_or_else(|| {
+                        FinalStateError::PosError("cycle overflow when feeding selector".into())
+                    })?)
+                    .map_err(|_| {
+                        FinalStateError::PosError("cycle overflow when feeding selector".into())
+                    })?;
```

Comment:
consider using helper functions and namings that help in self-documenting what is happening here.

Reply:
Done in refactor : 1405887a97bd4fbf5444e4b09f5cf2da48919a27